### DISCUSSION TITLE
ProtoPromise best practices?

### DISCUSSION THREAD

**Discussion by drew-512** [timcassell/ProtoPromise#189] [Q&A] [answered]
Hi!  Love ProtoPromise so far, thank you!

Long time veteran but only doing C# and Unity a couple years and currently wrapping my head around async/await. I'm trying get best practices in how ProtoPromise should play with `Task<T>`, `ValueTask<T>`  and `async`.  Assume Unity 2022.1 and later.

Suppose you're is building an awaitable suite from scratch using ProtoPromise -- where principle calls tend to be awaitable "all the way down" until they reach things like web requests, file loads, and embedded library calls. 

I would love to simplify and replace everything wholesale ProtoPromise but I find the whole `ValueTask<T>`  with `async` so friggin confusing.  It's tricky to understand what doesn't generate unneeded garbage -- any advice or convention suggestions are appreciated.  

Thanks ~ Drew

**Comment by timcassell** [comment, marked as answer]
Hi Drew,

ProtoPromise should be able to completely replace `Task` and `ValueTask` for all async methods that you write.

For any existing APIs that need to work with `Task` or `ValueTask` that you do not control, you can convert to/from ProtoPromise by using the `task.ToPromise()` extensions, or `promise.ToTask()`/`promise.AsValueTask()` extensions, as mentioned in the readme https://github.com/timcassell/ProtoPromise#task-interoperability.

> but I find the whole ValueTask<T> with async so friggin confusing

Basically, that behaves the same as `Task<T>`, but is more efficient if the method completed immediately. ProtoPromise has that same optimization built-in.

> It's tricky to understand what doesn't generate unneeded garbage

It's safe to assume `Task` and `ValueTask` will always generate garbage. There are some instances where they will not, but it's simpler to assume they will. ProtoPromise will never generate garbage*, unless you have object pooling disabled. *[Edit] See comment below.

Does that help?

**Reply by drew-512** [reply]
Beautiful!  Sorry I missed the info on interoperability as it was past all the advanced stuff.

I bet I'll have more questions in the coming week or two.

Thanks!

**Reply by timcassell** [reply]
Ah, that's useful feedback. I could move the advanced section to the bottom to help with the readability.

**Reply by timcassell** [reply]
Small correction - `Promise<T>.All` will generate garbage if you do not pass in an `IList` to be filled.

**Reply by drew-512** [reply]
As a data point, one of the parts I'm upgrading (if you agree it makes sense) is below, that manages requests than can be open for arbitrary amounts of time:

```.cs

public class HostSession : MonoBehaviour {

    ...
   
    void                                Update() {
        dispatchInbox(); // dispatch enqueued host Msgs on the main thread
     }

    public async Task                   PinAsync(PinOp op) {
        ... 
        var msg = Msg.Acquire();
        msg.Op = MsgOp.PinCell;
        msg.ValType = ValType.PinReq;
        msg.ValBuf = pb::MessageExtensions.ToByteString(op.PinReq);

        await queueNewReq(op.Target, msg);
    } 


    //// internal ////

    internal struct OpenReq {
        internal IOpenCell              origin;
        internal SemaphoreSlim          onComplete;
    }
    ulong                               _nextReqID = 1;
    Dictionary<ulong, OpenReq>          _openReqs = new Dictionary<ulong, OpenReq>();
    Dictionary<ulong, MsgBatch>         _fromHostInProgress = new Dictionary<ulong, MsgBatch>(); 
    ConcurrentQueue<MsgBatch>           _fromHost           = new ConcurrentQueue<MsgBatch>();

    Task                                queueNewReq(IOpenCell origin, Msg msg) {
        Task task = null;

        if (msg.Op != MsgOp.CloseReq) {
            msg.ReqID = _nextReqID++;
            var req = new OpenReq {
                origin = origin,
                onComplete = new SemaphoreSlim(0), // replace with Promise
            };
            task = req.onComplete.WaitAsync();
            _openReqs[msg.ReqID] = req;
        }
        _hostConn.EnqueueOutgoing(msg);
        
        return task;
    }

    
    // TODO: could this be remade with Promise so it dispatches via async/await?
    void                                dispatchInbox() {
        while (_fromHost.TryDequeue(out MsgBatch batch)) {                
            if (_openReqs.TryGetValue(batch.ReqID, out OpenReq req)) {
                req.origin.OnMsgBatch(batch);
                
                if (batch.ReqClosed) {
                    _openReqs.Remove(batch.ReqID);
                    req.onComplete?.Release();
                }
                
                batch.Reclaim();
            }
        }
    }

    // called by transport layer from a background thread to deliver msgs from the host
    internal void                       EnqueueIncoming(Msg msg) {
        _fromHostInProgress.TryGetValue(msg.ReqID, out MsgBatch batch);
        if (batch == null) {
            batch = MsgBatch.Acquire(msg.ReqID);
        }
        
        switch (msg.Op) {
            case MsgOp.CloseReq:
                batch.ReqClosed = true;
                _fromHostInProgress.Remove(batch.ReqID);
                _fromHost.Enqueue(batch);
                break;
            default:
                batch.Append(msg);
                break;
        }
    }
    
    void                                closeAllRequests() {
        MsgBatch closer = MsgBatch.MakeCloser();
        foreach (OpenReq req in _openReqs.Values) {
            req.origin.OnMsgBatch(closer);
            req.onComplete?.Release();
        }
        _openReqs.Clear();
    }

}
```

So based on your comments, I can replace `Task` wholesale and instead of `SemaphoreSlim` would just use `Promise.NewDeferred()`.  And instead of releasing the `SemaphoreSlim` to propagate a closed request, I would just call `deferredPromise.Resolve()` etc.

I'm also trying to grok if `dispatchInbox()` can be improved such that it uses async-await with Promise rather than the pattern of waiting for Unity's Update() etc.   But then that buys me also having to replace `ConcurrentQueue` with something that works with await etc (which I assume ProtoPromise could also support).

**Reply by timcassell** [reply]
It's a little weird to me that you use `SemaphoreSlim` like that (I would expect `TaskCompletionSource`), but yes, in that case you can replace it with `Promise.Deferred`. And yes you can replace `async Task` with `async Promise`

As for the queue and `Update` cycle, you don't actually need to do that, unless your incoming message processor is using a Unity API (which it looks like it's not). If you want to ensure that the callers of your API will continue on the main thread, rather than the background thread, you can use `await Promise.SwitchToForeground()` as the last line in your public function. That should eliminate the need for your `dispatchInbox()` function. Of course, if you do that, then you'll want to make sure the dictionaries are synchronized instead of the `ConcurrentQueue` (could use `ConcurrentDictionary`).

Or you can use `Promise.Run(() => { ... }, SynchronizationOption.Foreground)` to dispatch the message processing work to the main thread without relying on `Update`, then you won't need to synchronize the dictionaries at all (since all code will be ran on the main thread).

**Reply by drew-512** [reply]
Lovely, thank you.

On the `dispatchInbox` point, so since the data target (`req.origin`) is a `MonoBehaviour`, I'm hearing with the Dicts concurrent, I could just `SwitchToForeground()` do the dispatch, and switch back to the original background context.   I'm probably missing something, but perhaps a concern now is the invoking transport layer could get starved (though I suppose that would be happening anyway in that extreme case).  I suppose that's the tradeoff at hand.

**Reply by timcassell** [reply]
>  I could just `SwitchToForeground()` do the dispatch, and switch back to the original background context.

You could, but I'm not sure why you need to switch back to the background context there. Maybe I'm missing something, but it looks like you just need to notify the original caller that the operation is complete, so no more work needs to be done.

> perhaps a concern now is the invoking transport layer could get starved

How so?

**Reply by drew-512** [reply]
> > I could just `SwitchToForeground()` do the dispatch, and switch back to the original background context.
> 
> You could, but I'm not sure why you need to switch back to the background context there. Maybe I'm missing something, but it looks like you just need to notify the original caller that the operation is complete, so no more work needs to be done.

My thinking is that execution would dispatch on the main thread (now from within `EnqueueIncoming`) and then exit back into the network-based IO writer still the main thread (vs background thread).  Here's one of the reader implementations:
```.cs
        async void                          sessReader() {
        
            // Switch to a background
            await Await.BackgroundSyncContext();
            {
                var istream = _sess.ResponseStream;
                try {
                    bool ok = _isOpen;
                    while (ok) {
                        ok = await istream.MoveNext(_cancelToken);
                        if (ok) {
                            _client.EnqueueIncoming(istream.Current);
                            ok = _isOpen;
                        }
                    }
                } catch (grpc.RpcException ex) {
                    if (ex.StatusCode != grpc.StatusCode.Cancelled) {
                        await Await.UnitySyncContext();
                        Debug.Log(ex);
                    }
                }
                
                // Cancel the writer if the reader stops
                Cancel();
            }
```
Presumably, I would switch out the `Await.UnitySyncContext()` and `Await.BackgroundSyncContext` with Proto.Promise's.

(Basically, there's two possible transport impls set on startup: from an embedded DLL and the other via gRPC).

> 
> > perhaps a concern now is the invoking transport layer could get starved
> 
> How so?

The idea is that `sessReader()` can run outside the main thread while building up a `MsgBatch` (built from many Msgs).

**Reply by timcassell** [reply]
You don't have to completely exit the function in the background thread when you dispatch the main thread work. You can do some work in the background before or after you dispatch the main thread work. This does the same functionality as your code, but with ProtoPromise dispatching.

```cs
internal void EnqueueIncoming(Msg msg)
{
    // Background thread
    _fromHostInProgress.TryGetValue(msg.ReqID, out MsgBatch batch);
    if (batch == null)
    {
        batch = MsgBatch.Acquire(msg.ReqID);
    }

    if (msg.Op == MsgOp.CloseReq)
    {
        batch.ReqClosed = true;
        _fromHostInProgress.Remove(batch.ReqID);
        Promise.Run(() =>
        {
            // Main thread
            if (_openReqs.TryGetValue(batch.ReqID, out OpenReq req))
            {
                req.origin.OnMsgBatch(batch);

                if (batch.ReqClosed)
                {
                    _openReqs.Remove(batch.ReqID);
                    req.onComplete?.Release();
                }

                batch.Reclaim();
            }
        }, SynchronizationOption.Foreground)
            .Forget();
    }
    else
    {
        // Still on background thread
        batch.Append(msg);
    }
    
    // Do any more work you want while still on the background thread...
}
```

Or alternatively, like this:

```cs
internal void EnqueueIncoming(Msg msg)
{
    // Background thread
    _fromHostInProgress.TryGetValue(msg.ReqID, out MsgBatch batch);
    if (batch == null)
    {
        batch = MsgBatch.Acquire(msg.ReqID);
    }

    if (msg.Op == MsgOp.CloseReq)
    {
        batch.ReqClosed = true;
        _fromHostInProgress.Remove(batch.ReqID);
        Dispatch(batch).Forget();
    }
    else
    {
        // Still on background thread
        batch.Append(msg);
    }
    
    // Do any more work you want while still on the background thread...

    async Promise Dispatch(MsgBatch batch)
    {
        await Promise.SwitchToForeground();

        // Main thread
        if (_openReqs.TryGetValue(batch.ReqID, out OpenReq req))
        {
            req.origin.OnMsgBatch(batch);

            if (batch.ReqClosed)
            {
                _openReqs.Remove(batch.ReqID);
                req.onComplete?.Release();
            }

            batch.Reclaim();
        }
    }
}
```

**Reply by timcassell** [reply]
> Presumably, I would switch out the `Await.UnitySyncContext()` and `Await.BackgroundSyncContext` with Proto.Promise's.

You can, but you don't have to. Since it's async/await, those should work just fine with `async Promise`.

**Reply by drew-512** [reply]
Oh perfect, thanks for breaking that down my good man.   Happy to adopt ProtoPromise wholesale now that I'm understanding the idioms.  

Been heads down on an all-platform, all-purpose AV Player in Unity -- think VLC meets Winamp/iTunes with high fidelity visuals and UX.  It has more (a Go infra plugin system that can drive the UI) but I don't want to be a bore.  Looking fwd to crediting ProtoPrimise and will shoutout in a few weeks when the alpha is a little less rough.   If you've ever seen the original iTunes visualizer (G-Force) then you've seen some of my old stuff.   Happy to connect further if that's interesting.

**Reply by timcassell** [reply]
Great, happy to help! I'll look forward to it. :smiley:

**Reply by drew-512** [reply]
Digging on PP!   

Quick question on chaining conventions and efficiency...

Suppose we start with:
```.cs
    
    Promise<bool>                doRequirement() {            
        ...
    }
    
    Promise<bool>                doThing() {            
        ...
    }
    
    public async bool            DoThing() {            
        bool ok = true;
        
        if (someRequirement()) 
            ok = await doRequirement();
        
        if (ok)
            ok = await doThing();
        
        return ok;
    }
```

But since it's nice to have a version of DoThing() that can be used with `.Forget()` (rather than the caller having to make it an `async` call), what does the non-async version look like? 


```.cs
    public Promise<bool>         DoThing() {            
        Promise<bool> p = Promise<bool>.Resolved(true);
        
        if (someRequirement()) 
            p = doRequirement();
        
        p = p.Then((bool ok) => {
            if (ok)
                return doThing();
            else
                return Promise<bool>.Resolved(false);
        });
        
        return p;
    }
```

I wanted to check if this as clean as it gets in terms of code readability and GC load before I start taking on larger blocks of code.

**Reply by timcassell** [reply]
`.Then` really exists for old runtimes before `async`/`await` was available. Tbh, `async`/`await` is much more readable to me.

As for which API is more efficient, it really depends on your use-case. You would have to benchmark it, and in most likelihood it will be a negligible difference. `.Then` uses a bit less memory if you only need to append a single callback, but if you need to chain multiple `.Then`s together, `async`/`await` is lazier, so you might end up using less memory that way. But it's not always the case, so you would need to measure it yourself. Also, the memory used is pooled either way by default, so it may be difficult to measure.

1 advantage that `.Then` has is it automatically normalizes progress for you, while you need to normalize it manually with `async`/`await` if you use it.

I would default to `async`/`await`.

> But since it's nice to have a version of DoThing() that can be used with .Forget() (rather than the caller having to make it an async call)

I'm not sure what you mean by that. Either your function is declared as `async Promise<bool>` and you use `await`s, or it's declared as `Promise<bool>` and you use `.Then`. it's returning the same type either way, and the caller will consume it the same either way.

Also, there's a bug in your code. `Promise<bool> p = Promise<bool>.Resolved(true);` is not forgotten if `someRequirement()` returns `true`. You should forget it, then assign the new promise (or better, assign it in the else clause).

**Reply by timcassell** [reply]
Also, it's easy to accidentally allocate garbage through closures by using the `.Then` API. `async`/`await` doesn't use closures, it stores all the state in a backing state machine, so that's not something you need to worry about.

**Reply by drew-512** [reply]
Awesome, thanks.   I think I'm starting to get the hang of it. 

As a relative newcomer to C#, I've been reflecting on why it took me so long to really get the ins and out of async await.   I think it's been a combo of noise (tons of people who write or post on stack overflow that really shouldn't) and excessively voluminous enterprisey docs such that ya get a lot of fluff and specs but rarely no "this is what's happening under the hood" as a way to help understand and use it effectively.  That's why I've been such a Go fan since since my C++ days -- the language and runtime spec is purposefully small.  

Anyway, thanks again for ProtoPromise as its made async await enjoyable to use in Unity!

**Reply by timcassell** [reply]
Ha, yeah. People tend to complicate async/await, but it really is as simple as, "pause this function, and continue after the awaited task is complete". Other more complex things should come after getting that basic idea down.

> but rarely no "this is what's happening under the hood"

Believe me, I did a ton of research while writing this library, and at times I just had to look right at the source code to see what the heck MS was doing with tasks. Their APIs are ridiculously complex. I think I simplified things enough to be easily approachable, while still providing the tools necessary to do more complex things when necessary.

> Anyway, thanks again for ProtoPromise as its made async await enjoyable to use in Unity!

I'm glad to hear it! :heart:

**Reply by drew-512** [reply]
Exactly -- you expose the perfect amount of complexity (yielding simplicity).  And in rare form, C# type-safety and conversion does the rest without adding more hoops. :D  That you also bench and QA against Unity is huge and underrated.  ProtoPromise is literally a prayer answered. 

So many C# packages either fall down in Unity, are bloated (Google.Protobufs), or have painful dependencies (e.g. `Microsoft.Logging` etc).    An example of that last one, I've been wanting to switch to [Capn Proto](https://capnproto.org/) but I don't have the C# chops to get it the [C# version](https://github.com/c80k/capnproto-dotnetcore) of it building and linking under Unity.   As this prototype hits more milestones and funding is unlocked, my plan was to use this [Unity friendly impl](https://github.com/kosek-com/capnproto-unity) of CapnProto or and/or sponsor development for it.  Now I want to see if it will integrate well with PP.
